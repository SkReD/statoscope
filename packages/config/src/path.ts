import path from 'path';
import module from 'module';
import chalk from 'chalk';

export enum PackageAliasType {
  PLUGIN = 'PLUGIN',
  REPORTER = 'REPORTER',
}

export function normalizePath(source: string, rootDir: string): string {
  return source.replace('<rootDir>', rootDir);
}

export function makeRequireFromPath(pathname: string): NodeRequire {
  return module.createRequire(path.resolve(pathname, '_'));
}

// package namespace (empty string when not set) followed by the package name (wo namespace)
type PackageName = [string, string];

const packageAliasPrefixes = new Map<PackageAliasType, Record<PrefixType, string>>();

enum PrefixType {
  STATOSCOPE_NS = '@statoscope',
  NON_STATOSCOPE_NS = '@NON_STATOSCOPE_NS',
}

packageAliasPrefixes.set(PackageAliasType.PLUGIN, {
  [PrefixType.STATOSCOPE_NS]: 'stats-validator-plugin',
  [PrefixType.NON_STATOSCOPE_NS]: 'statoscope-stats-validator-plugin',
});

packageAliasPrefixes.set(PackageAliasType.REPORTER, {
  [PrefixType.STATOSCOPE_NS]: 'stats-validator-reporter',
  [PrefixType.NON_STATOSCOPE_NS]: 'statoscope-stats-validator-reporter',
});

/**
 * Concatenate namespace (when provided) with the package name.
 */
function getFullPackageName([namespace, name]: PackageName): string {
  if (namespace) {
    return `${namespace}/${name}`;
  }

  return name;
}

/**
 * Return error message based on the list of package aliases (provided by the user and alternative).
 */
function getAliasPackageResolutionError(packageAliases: PackageName[]): string {
  const providedAlias = getFullPackageName(packageAliases[0]);

  let errorMessage = `Can't resolve package ${chalk.yellow.italic(providedAlias)}.\n`;

  const { italic: italicChalk } = chalk;

  if (packageAliases.length > 1) {
    const alternativeAlias = getFullPackageName(packageAliases[1]);

    errorMessage += `Also tried to resolve it with ${italicChalk.yellow(
      alternativeAlias,
    )} alias which didn't work either.\n\n`;
  } else {
    errorMessage += '\n\n';
  }

  errorMessage += 'Try installing the package locally:\n';
  errorMessage += `- with ${italicChalk('npm')}: ${chalk.bgGrey(
    'npm i -D ' + providedAlias,
  )} (or corresponding package alias)\n`;
  errorMessage += `- with ${italicChalk('yarn')}: ${chalk.bgGrey(
    'yarn add -D ' + providedAlias,
  )} (or corresponding package alias)\n`;

  return errorMessage;
}

/**
 * Resolve full or short package name to it's absolute path.
 * @param packageAliasType - Plugin or reporter enum.
 * @param aliasName - Package name (alias). Can be passed in the "full" or "short" form.
 *   Will append prefixes based on packageAliasType value in order to resolve the short form alias.
 * @param fromDir - Directory used as "root" while resolving the package.
 */
export function resolveAliasPackage(
  packageAliasType: PackageAliasType,
  aliasName: string,
  fromDir: string,
): string {
  const localRequire = makeRequireFromPath(fromDir);
  aliasName = normalizePath(aliasName, fromDir);

  if (aliasName.startsWith('.') || path.isAbsolute(aliasName)) {
    localRequire(aliasName);
    return aliasName;
  }

  let packageNamespace = '';
  let packageName = aliasName;

  const packageNameRegex = /^(@.+?)\/(.+)/;

  if (packageNameRegex.test(aliasName)) {
    [, packageNamespace = '', packageName] = aliasName.match(packageNameRegex)!;
  }

  // can have from one up to two entries
  const packageAliases: PackageName[] = [
    [packageNamespace, packageName], // original form
  ];

  const prefixes = packageAliasPrefixes.get(packageAliasType)!;

  const prefix =
    packageNamespace === PrefixType.STATOSCOPE_NS
      ? prefixes[PrefixType.STATOSCOPE_NS]
      : prefixes[PrefixType.NON_STATOSCOPE_NS];

  if (!packageName.startsWith(prefix)) {
    packageAliases.push([packageNamespace, `${prefix}-${packageName}`]);
  }

  const paths = packageAliases.map(([namespace, packageName]) =>
    path.join(namespace, packageName),
  );

  for (const path of paths) {
    try {
      localRequire(path);
      return path;
      // eslint-disable-next-line no-empty
    } catch (e) {}
  }

  throw new Error(getAliasPackageResolutionError(packageAliases));
}
