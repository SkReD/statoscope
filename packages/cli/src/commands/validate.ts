import path from 'path';
import { Argv } from 'yargs';
import Validator from '@statoscope/stats-validator';
import ConsoleReporter from '@statoscope/stats-validator-reporter-console';
import { Reporter } from '@statoscope/types/types/validation/reporter';
import { makeReporterInstance } from '@statoscope/config/dist/make-reporter-instance';
import { requireConfig } from '@statoscope/config';
import legacyWebpackValidator from './legacyWebpackValidator';

export default function (yargs: Argv): Argv {
  return yargs.command(
    'validate [validator] [input]',
    `Validate or compare JSON stats
Examples:
Validate stats: validate --input path/to/stats.json
Compare stats: validate --input path/to/branch.json --reference path/to/master.json
Validate stats [DEPRECATED]: validate --validator path/to/validator.js --input path/to/stats.json
Validate multiple stats [DEPRECATED]: validate --validator path/to/validator.js --input path/to/stats.json --input path/to/another/stats.json`,
    (yargs) => {
      return (
        yargs
          // todo remove in 6.0
          .option('validator', {
            describe: '[DEPRECATED] path to validator script',
            alias: 'v',
            type: 'string',
          })
          .option('config', {
            describe: 'path to statoscope config',
            alias: 'c',
            type: 'string',
          })
          .option('input', {
            describe: 'path to a stats.json',
            alias: 'i',
            type: 'string',
          })
          .option('reference', {
            describe: 'path to a stats-file to compare with',
            alias: 'r',
            type: 'string',
          })
          .option('warn-as-error', {
            describe: 'treat warn as error',
            alias: 'w',
            type: 'boolean',
          })
          .demandOption(['input'])
          .array('input')
      );
    },
    async (argv) => {
      if (argv.validator) {
        await legacyWebpackValidator({
          validator: argv.validator,
          input: argv.input,
          warnAsError: argv['warn-as-error'],
        });
        return;
      }

      const { path: configPath, config } = requireConfig(argv.config);
      const rootPath = path.dirname(configPath);
      const validateConfig = config.validate ?? { rules: {} };

      if (argv['warn-as-error']) {
        validateConfig.warnAsError = true;
      }

      const validator = new Validator(validateConfig, rootPath);
      const reporters: Reporter[] = [];

      if (config?.silent !== true) {
        if (config?.validate?.reporters) {
          for (const item of config.validate.reporters) {
            reporters.push(makeReporterInstance(item, rootPath));
          }
        } else {
          reporters.push(new ConsoleReporter());
        }
      }

      const result = await validator.validate(
        path.resolve(argv.input[0]),
        argv.reference ? path.resolve(argv.reference) : void 0,
      );

      for (const rule of result.rules) {
        const storage = rule.api.getStorage();

        if (rule.execParams.mode === 'error' && storage.length) {
          process.exitCode = 1;
          break;
        }
      }

      for (const reporter of reporters) {
        await reporter.run(result);
      }
    },
  );
}
