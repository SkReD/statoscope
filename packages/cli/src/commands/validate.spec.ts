import os from 'os';
import path from 'path';
import yargs from 'yargs';
import validate from './validate';

const configFixturePath = path.resolve(
  __dirname,
  '../../../../test/fixtures/cli/validate/config.js',
);

const configWithReportersFixturePath = path.resolve(
  __dirname,
  '../../../../test/fixtures/cli/validate/config-with-reporters.js',
);

const configReferenceFixturePath = path.resolve(
  __dirname,
  '../../../../test/fixtures/cli/validate/config-reference.js',
);

const inputFixturePath = path.resolve(
  __dirname,
  '../../../../test/bundles/v5/simple/stats-dev.json',
);

const referenceFixturePath = path.resolve(
  __dirname,
  '../../../../test/bundles/v4/simple/stats-dev.json',
);

const output: string[][] = [];
let consoleLogSpy: jest.SpyInstance<void, string[]> | null = null;

beforeEach(() => {
  consoleLogSpy = jest
    .spyOn(global.console, 'log')
    .mockImplementation((...args: string[]) => {
      output.push(
        args.map((o) =>
          String(o).replace(process.cwd(), '<pwd>').replace(os.tmpdir(), '<tmp>'),
        ),
      );
    });
});

afterEach(() => {
  consoleLogSpy?.mockRestore();
  output.length = 0;
});

function getOutput(): string[][] {
  return output;
}

describe('with config', () => {
  describe('explicit', () => {
    test('single', async () => {
      let y = yargs([
        'validate',
        '--config',
        configFixturePath,
        '--input',
        inputFixturePath,
      ]);

      y = validate(y);
      y.fail((_, error) => {
        console.error(error);
      });

      await y.argv;

      expect(getOutput()).toMatchSnapshot();
    });

    test('reference', async () => {
      let y = yargs([
        'validate',
        '--config',
        configReferenceFixturePath,
        '--input',
        inputFixturePath,
        '--reference',
        referenceFixturePath,
      ]);

      y = validate(y);
      y.fail((_, error) => {
        console.error(error);
      });

      await y.argv;

      expect(getOutput()).toMatchSnapshot();
    });

    test('reporter', async () => {
      let y = yargs([
        'validate',
        '--config',
        configWithReportersFixturePath,
        '--input',
        inputFixturePath,
      ]);

      y = validate(y);
      y.fail((_, error) => {
        console.error(error);
      });

      await y.argv;

      expect(getOutput()).toMatchSnapshot();
    });
  });
});
