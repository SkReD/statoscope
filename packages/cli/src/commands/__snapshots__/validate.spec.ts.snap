// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with config explicit reference 1`] = `
[
  [
    "input.json",
  ],
  [
    "ccac20dc279ee29ab60e",
  ],
  [
    "error  Module ./simple/src/index.ts should not be used  webpack/restricted-modules",
  ],
  [
    "error  Module ./simple/src/index2.ts should not be used  webpack/restricted-modules",
  ],
  [
    "error  Module ./simple/src/sub-module.ts should not be used  webpack/restricted-modules",
  ],
  [
    "error  Module ./simple/src/index2.css should not be used  webpack/restricted-modules",
  ],
  [
    "error  Module ./simple/src/test.txt should not be used  webpack/restricted-modules",
  ],
  [
    "error  Module ./simple/src/dyn.ts should not be used  webpack/restricted-modules",
  ],
  [
    "error  Module ./simple/src/statoscope.png should not be used  webpack/restricted-modules",
  ],
  [
    "error  foo@1.0.0 should not be used  webpack/restricted-packages",
  ],
  [
    "Instances:",
  ],
  [
    "- simple/node_modules/foo  1.0.0",
  ],
  [
    "unknown compilation",
  ],
  [
    "error  Usage of ./node_modules/is-array/index.js was increased from 2 to 4  webpack/diff-deprecated-modules",
  ],
  [
    "error  Usage of external "extLib" was increased from 0 to 1  webpack/diff-deprecated-modules",
  ],
  [
    "error  Usage of ./simple/src/test.txt was increased from 0 to 1  webpack/diff-deprecated-modules",
  ],
  [
    "error  Usage of ./simple/src/statoscope.png was increased from 0 to 2  webpack/diff-deprecated-modules",
  ],
  [
    "error  Usage of is-array was increased from 2 to 4  webpack/diff-deprecated-packages",
  ],
  [
    "",
  ],
  [
    "13 problems (13 errors, 0 warnings)",
  ],
]
`;

exports[`with config explicit reporter 1`] = `
[
  [
    "undefined",
  ],
  [
    "Module ./simple/src/index.ts should not be used",
  ],
  [
    "Module ./simple/src/index2.ts should not be used",
  ],
  [
    "Module ./simple/src/sub-module.ts should not be used",
  ],
  [
    "Module ./simple/src/index2.css should not be used",
  ],
  [
    "Module ./simple/src/test.txt should not be used",
  ],
  [
    "Module ./simple/src/dyn.ts should not be used",
  ],
  [
    "Module ./simple/src/statoscope.png should not be used",
  ],
  [
    "foo@1.0.0 should not be used",
  ],
]
`;

exports[`with config explicit single 1`] = `
[
  [
    "input.json",
  ],
  [
    "ccac20dc279ee29ab60e",
  ],
  [
    "error  Module ./simple/src/index.ts should not be used  webpack/restricted-modules",
  ],
  [
    "error  Module ./simple/src/index2.ts should not be used  webpack/restricted-modules",
  ],
  [
    "error  Module ./simple/src/sub-module.ts should not be used  webpack/restricted-modules",
  ],
  [
    "error  Module ./simple/src/index2.css should not be used  webpack/restricted-modules",
  ],
  [
    "error  Module ./simple/src/test.txt should not be used  webpack/restricted-modules",
  ],
  [
    "error  Module ./simple/src/dyn.ts should not be used  webpack/restricted-modules",
  ],
  [
    "error  Module ./simple/src/statoscope.png should not be used  webpack/restricted-modules",
  ],
  [
    "error  foo@1.0.0 should not be used  webpack/restricted-packages",
  ],
  [
    "Instances:",
  ],
  [
    "- simple/node_modules/foo  1.0.0",
  ],
  [
    "",
  ],
  [
    "8 problems (8 errors, 0 warnings)",
  ],
]
`;
